src/lsp.ts

```diff
@@ -0,0 +1,59 @@
+import * as which from "which";
+import {
+  canDownloadLanguageServer,
+  findLanguageServer,
+  setupLanguageServer,
+} from "./utils";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { LanguageClient } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  config: WorkspaceConfiguration
+  ls: LanguageClient | null = null
+  executableName: string
+  context: ExtensionContext
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName
+    this.context = context
+    this.config = workspace.getConfiguration("mesonbuild");
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
+  }
+
+  abstract startLanguageServer(serverModule: string): void
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
+    this.dispose();
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer(serverModule);
+  }
+
+  canDownloadLanguageServer(): boolean {
+    return canDownloadLanguageServer(this)
+  }
+
+  get languageServerPath(): string | null {
+    return findLanguageServer(this) || this.config.languageServerPath || which.sync(this.executableName, { nothrow: true })
+  }
+
+  async setupLanguageServer() {
+    setupLanguageServer(this)
+  }
+
+  dispose() {
```

Comment:
It has been a while since I wrote TS. Do you have to mark this private?

Reply:
Not sure, but making it private breaks nothing

Reply:
I only say that because it is only called in finalize.